File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
      vector<vector<string>> ans;
      vector<string> temp;
      int n = s.size();
      // pre compute if string [i--j] is palindrome or not
      vector<vector<bool>>isPalindrome(n,vector<bool>(n,true));

      // to check
      //Build palindrome lookup table using dynamic programming
        // For any substring, it's a palindrome if:
        // 1. First and last characters match AND
        // 2. The substring between them is also a palindrome
      for(int i= n-1;i>-1;i--)
      {
        for(int j=i+1;j<n;j++)
        {
            isPalindrome[i][j]= (s[i]==s[j])&& isPalindrome[i+1][j-1];
        }
      }
      Helper(0,isPalindrome, s,ans,temp);  
      return ans;
        
    }
    void Helper(int index,vector<vector<bool>>& isPalindrome, string s, vector<vector<string>>& ans, vector<string>& temp)
    {
       // cout<<"called"<<endl;
        if(index==s.size()){
            ans.push_back(temp);
            return;
        }
        for(int i = index ; i < s.size();i++){
            //cout<<"insideloop"<<endl;
            string a = s.substr(index,i-index+1);
            if(isPalindrome[index][i]==true){
                temp.push_back(a);
                Helper(i+1,isPalindrome,s,ans,temp);
                temp.pop_back();
            }
            
        }
    } 
    
};

File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        List<String> temp = new ArrayList<>();
        Helper(s,0,temp,ans);
        return ans;
    }
    void Helper(String s,int index,List<String> temp,List<List<String>> ans)
    {
        if(index==s.length()) {
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = index+1;i<=s.length();i++)
        {
            String a = s.substring(index,i);
            if(isPalindrome(a)==true){
                temp.add(a);
                Helper(s,i,temp,ans);
                temp.remove(temp.size()-1);
            }
        }
    }
    boolean isPalindrome(String s)
    {
        int i = 0;
        int j = s.length()-1;
        while(i<j)
        {
            if(s.charAt(i)!=s.charAt(j)) return false;
            i++;
            j--;
        }
        return true;
    }
}